source/validate_capability.cpp

```diff
@@ -0,0 +1,119 @@
+
+#include "diagnostic.h"
+#include "opcode.h"
+#include "val/instruction.h"
+#include "val/validation_state.h"
+
+namespace libspirv {
+
+namespace {
+
+bool IsSupportGuaranteedVulkan_1_0(uint32_t capability) {
+  switch (capability) {
+    case SpvCapabilityMatrix:
+    case SpvCapabilityShader:
+    case SpvCapabilityInputAttachment:
+    case SpvCapabilitySampled1D:
+    case SpvCapabilityImage1D:
+    case SpvCapabilitySampledBuffer:
+    case SpvCapabilityImageBuffer:
+    case SpvCapabilityImageQuery:
+    case SpvCapabilityDerivativeControl:
+      return true;
+  }
+  return false;
+}
+
+bool IsSupportOptionalVulkan_1_0(uint32_t capability) {
+  switch (capability) {
+    case SpvCapabilityGeometry:
+    case SpvCapabilityTessellation:
+    case SpvCapabilityFloat64:
+    case SpvCapabilityInt64:
+    case SpvCapabilityInt16:
+    case SpvCapabilityTessellationPointSize:
+    case SpvCapabilityGeometryPointSize:
+    case SpvCapabilityImageGatherExtended:
+    case SpvCapabilityStorageImageMultisample:
+    case SpvCapabilityUniformBufferArrayDynamicIndexing:
+    case SpvCapabilitySampledImageArrayDynamicIndexing:
+    case SpvCapabilityStorageBufferArrayDynamicIndexing:
+    case SpvCapabilityStorageImageArrayDynamicIndexing:
+    case SpvCapabilityClipDistance:
+    case SpvCapabilityCullDistance:
+    case SpvCapabilityImageCubeArray:
+    case SpvCapabilitySampleRateShading:
+    case SpvCapabilitySparseResidency:
+    case SpvCapabilityMinLod:
+    case SpvCapabilitySampledCubeArray:
+    case SpvCapabilityImageMSArray:
+    case SpvCapabilityStorageImageExtendedFormats:
+    case SpvCapabilityInterpolationFunction:
+    case SpvCapabilityStorageImageReadWithoutFormat:
+    case SpvCapabilityStorageImageWriteWithoutFormat:
+    case SpvCapabilityMultiViewport:
+      return true;
+  }
+  return false;
+}
+
+}  // namespace
+
+// Validates that capability declarations use operands allowed in the current
+// context.
+spv_result_t CapabilityPass(ValidationState_t& _,
+                           const spv_parsed_instruction_t* inst) {
+  const SpvOp opcode = static_cast<SpvOp>(inst->opcode);
+  if (opcode != SpvOpCapability)
+    return SPV_SUCCESS;
+
+  if (inst->num_operands != 1) {
+    return _.diag(SPV_ERROR_INVALID_CAPABILITY)
+        << "OpCapability must have exactly one operand.";
+  }
+
+  const spv_parsed_operand_t& operand = inst->operands[0];
+
+  if (operand.num_words != 1) {
+    return _.diag(SPV_ERROR_INVALID_CAPABILITY)
+        << "OpCapability operand must consist of exactly one word.";
+  }
+
+  assert(operand.offset < inst->num_words);
+
+  const uint32_t capability = inst->words[operand.offset];
+
+  const auto env = _.context()->target_env;
+  if (env == SPV_ENV_VULKAN_1_0) {
+    if (!IsSupportGuaranteedVulkan_1_0(capability) &&
+        !IsSupportOptionalVulkan_1_0(capability)) {
+      return _.diag(SPV_ERROR_INVALID_CAPABILITY)
+          << "Capability operand not allowed by Vulkan specification"
```

Comment:
Please add unit test/tests that exercise this code and checks for this message.
you can use `getDiagnosticString()` to get the message in the unit tests (see existing tests)